File: obel/obeltests/coord.m
//
//  coord.m
//  obeltests
//
//  Created by Jack Mousseau on 10/2/21.
//

#import <obel/coord.h>
#import <XCTest/XCTest.h>

@interface ObelCoordTests : XCTestCase

@end

@implementation ObelCoordTests

- (void)testCoordLayout {
    XCTAssertEqual(sizeof(obel_coord_t), sizeof(uint64_t));
}

- (void)testMinHeap {
    obel_coord_t coords[6] = { { .bits = 0 }, { 0 }, { 0 }, { 0 }, { 0 }, { 0 } };
    obel_coord_heap_t min_heap = {
        .order = obel_heap_order_min,
        .mask = 0xffff,
        .capacity = 5,
        .count = 0,
        .coords = coords
    };

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 3 }), 1);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 3);

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 4 }), 2);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 3);

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 2 }), 1);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 2);

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 5 }), 4);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 2);

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 1 }), 1);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 1);

    XCTAssertEqual(obel_coord_heap_insert(&min_heap, (obel_coord_t){ .bits = 0 }), 0);
    XCTAssertEqual(min_heap.coords[OBEL_HEAP_ROOT].bits, 1);

    XCTAssertEqual(obel_coord_heap_remove(&min_heap, OBEL_HEAP_ROOT).bits, 1);
    XCTAssertEqual(obel_coord_heap_remove(&min_heap, OBEL_HEAP_ROOT).bits, 2);
    XCTAssertEqual(obel_coord_heap_remove(&min_heap, OBEL_HEAP_ROOT).bits, 3);
    XCTAssertEqual(obel_coord_heap_remove(&min_heap, OBEL_HEAP_ROOT).bits, 4);
    XCTAssertEqual(obel_coord_heap_remove(&min_heap, OBEL_HEAP_ROOT).bits, 5);
}

- (void)testMaxHeap {
    obel_coord_t coords[6] = { { .bits = UINT16_MAX }, { 0 }, { 0 }, { 0 }, { 0 }, { 0 } };
    obel_coord_heap_t max_heap = {
        .order = obel_heap_order_max,
        .mask = 0xffff,
        .capacity = 5,
        .count = 0,
        .coords = coords
    };

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 1 }), 1);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 1);

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 2 }), 1);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 2);

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 4 }), 1);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 4);

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 5 }), 1);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 5);

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 3 }), 5);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 5);

    XCTAssertEqual(obel_coord_heap_insert(&max_heap, (obel_coord_t){ .bits = 6 }), 0);
    XCTAssertEqual(max_heap.coords[OBEL_HEAP_ROOT].bits, 5);

    XCTAssertEqual(obel_coord_heap_remove(&max_heap, OBEL_HEAP_ROOT).bits, 5);
    XCTAssertEqual(obel_coord_heap_remove(&max_heap, OBEL_HEAP_ROOT).bits, 4);
    XCTAssertEqual(obel_coord_heap_remove(&max_heap, OBEL_HEAP_ROOT).bits, 3);
    XCTAssertEqual(obel_coord_heap_remove(&max_heap, OBEL_HEAP_ROOT).bits, 2);
    XCTAssertEqual(obel_coord_heap_remove(&max_heap, OBEL_HEAP_ROOT).bits, 1);
}

- (void)testCoordTable {
    const uint32_t capacity = 5;
    obel_coord_t coords[capacity] = { 0 };
    obel_coord_table_t table = {
        .count = 0,
        .capacity = capacity,
        .coords = coords
    };

    obel_coord_t a = obel_coord_make(1, 2, 3);
    obel_coord_t b = obel_coord_make(2, 3, 4);
    obel_coord_t c = obel_coord_make(3, 4, 5);
    obel_coord_t d = obel_coord_make(4, 5, 6);
    obel_coord_t e = obel_coord_make(5, 6, 7);
    obel_coord_t f = obel_coord_make(6, 7, 8);

    {
        uint32_t upper = (capacity + 1);
        uint32_t i_a, i_b, i_c, i_d, i_e, i_f = upper;
        XCTAssertTrue(obel_coord_table_insert(&table, a, &i_a) && i_a < upper && table.count == 1);
        XCTAssertTrue(obel_coord_table_insert(&table, b, &i_b) && i_b < upper && table.count == 2);
        XCTAssertTrue(obel_coord_table_insert(&table, c, &i_c) && i_c < upper && table.count == 3);
        XCTAssertTrue(obel_coord_table_insert(&table, d, &i_d) && i_d < upper && table.count == 4);
        XCTAssertTrue(obel_coord_table_insert(&table, e, &i_e) && i_e < upper && table.count == 5);
        XCTAssertTrue(!obel_coord_table_insert(&table, f, &i_f) && i_f == upper && table.count == 5);

        uint32_t j_a, j_b, j_c, j_d, j_e, j_f = upper;
        XCTAssertTrue(obel_coord_table_find(&table, a, &j_a) && j_a == i_a && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, b, &j_b) && j_b == i_b && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, c, &j_c) && j_c == i_c && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, d, &j_d) && j_d == i_d && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, e, &j_e) && j_e == i_e && table.count == 5);
        XCTAssertTrue(!obel_coord_table_find(&table, f, &j_f) && j_f == upper && table.count == 5);

        uint32_t k_a, k_b, k_c, k_d, k_e, k_f = upper;
        XCTAssertTrue(obel_coord_table_remove(&table, a, &k_a) && k_a == i_a && table.count == 4);
        XCTAssertTrue(obel_coord_table_remove(&table, b, &k_b) && k_b == i_b && table.count == 3);
        XCTAssertTrue(obel_coord_table_remove(&table, c, &k_c) && k_c == i_c && table.count == 2);
        XCTAssertTrue(obel_coord_table_remove(&table, d, &k_d) && k_d == i_d && table.count == 1);
        XCTAssertTrue(obel_coord_table_remove(&table, e, &k_e) && k_e == i_e && table.count == 0);
        XCTAssertTrue(!obel_coord_table_remove(&table, f, &k_f) && k_f == upper && table.count == 0);
    }

    {
        XCTAssertTrue(obel_coord_table_insert(&table, a) && table.count == 1);
        XCTAssertTrue(obel_coord_table_insert(&table, b) && table.count == 2);
        XCTAssertTrue(obel_coord_table_insert(&table, c) && table.count == 3);
        XCTAssertTrue(obel_coord_table_insert(&table, d) && table.count == 4);
        XCTAssertTrue(obel_coord_table_insert(&table, e) && table.count == 5);
        XCTAssertTrue(!obel_coord_table_insert(&table, f) && table.count == 5);

        XCTAssertTrue(obel_coord_table_find(&table, a) && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, b) && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, c) && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, d) && table.count == 5);
        XCTAssertTrue(obel_coord_table_find(&table, e) && table.count == 5);
        XCTAssertTrue(!obel_coord_table_find(&table, f) && table.count == 5);

        XCTAssertTrue(obel_coord_table_remove(&table, a) && table.count == 4);
        XCTAssertTrue(obel_coord_table_remove(&table, b) && table.count == 3);
        XCTAssertTrue(obel_coord_table_remove(&table, c) && table.count == 2);
        XCTAssertTrue(obel_coord_table_remove(&table, d) && table.count == 1);
        XCTAssertTrue(obel_coord_table_remove(&table, e) && table.count == 0);
        XCTAssertTrue(!obel_coord_table_remove(&table, f) && table.count == 0);
    }
}

- (void)testCoordGridBin {
    float stride = 1.5;
    obel_vector3_t origin = obel_vector3_make(1.f, 2.f, 3.f);
    obel_vector3_t point = obel_vector3_make(9.f, 5.f, 4.f);

    {
        obel_coord_t coord = obel_coord_grid_bin(stride, origin, point);
        XCTAssertEqual(coord.x, 5);
        XCTAssertEqual(coord.y, 2);
        XCTAssertEqual(coord.z, 0);
    }

    {
        obel_coord_grid_t grid = {
            .origin = origin,
            .stride = stride
        };

        obel_coord_t coord = obel_coord_grid_bin(&grid, point);
        XCTAssertEqual(coord.x, 5);
        XCTAssertEqual(coord.y, 2);
        XCTAssertEqual(coord.z, 0);
    }
}

@end

File: obel/obel/coord.c
//
//  coord.c
//  obel
//
//  Created by Jack Mousseau on 10/2/21.
//

#include <obel/coord.h>

// MARK: - Sparse Coordinate Table

OBEL_COORD_FUNC uint64_t obel_coord_xyz_hash(obel_coord_t coord) {
    uint64_t hash = ((int64_t)coord.x << 48);
    hash |= ((uint64_t)coord.y << 32);
    hash |= ((uint64_t)coord.z << 16);
    return (OBEL_COORD_BITS_OCCUPIED | (hash & 0x0000ffffffffffff));
}

OBEL_COORD_FUNC bool obel_coord_table_insert(obel_coord_table_t *table, obel_coord_t coord) {
    return obel_coord_table_insert(table, coord, NULL);
}

OBEL_COORD_FUNC bool obel_coord_table_insert(obel_coord_table_t *table, obel_coord_t coord, uint32_t *i) {
    if (table->count == table->capacity) {
        return false;
    }

    uint32_t found_i;
    obel_coord_table_find(table, coord, &found_i);
    table->coords[found_i] = coord;
    table->coords[found_i].bits |= OBEL_COORD_BITS_OCCUPIED;
    table->count++;
    OBEL_SAFE_DEREF_ASSIGN(i, found_i);
    return true;
}

OBEL_COORD_FUNC bool obel_coord_table_remove(obel_coord_table_t *table, obel_coord_t coord) {
    return obel_coord_table_remove(table, coord, NULL);
}

OBEL_COORD_FUNC bool obel_coord_table_remove(obel_coord_table_t *table, obel_coord_t coord, uint32_t *i) {
    uint32_t found_i;
    if (obel_coord_table_find(table, coord, &found_i)) {
        table->coords[found_i].bits &= ~OBEL_COORD_BITS_OCCUPIED;
        table->count--;
        OBEL_SAFE_DEREF_ASSIGN(i, found_i);
        return true;
    } else {
        return false;
    }
}

OBEL_COORD_FUNC bool obel_coord_table_find(obel_coord_table_t *table, obel_coord_t coord) {
    return obel_coord_table_find(table, coord, NULL);
}

OBEL_COORD_FUNC bool obel_coord_table_find(obel_coord_table_t *table, obel_coord_t coord, uint32_t *i) {
    uint32_t current = (uint32_t)(obel_coord_xyz_hash(coord) % (uint64_t)table->capacity);
    uint32_t end = current == 0 ? table->capacity - 1 : current - 1;

    obel_coord_t existing = { 0 };
    while (current != end && (existing = table->coords[current]).bits & OBEL_COORD_BITS_OCCUPIED) {
        if (existing.x == coord.x && existing.y == coord.y && existing.z == coord.z) {
            OBEL_SAFE_DEREF_ASSIGN(i, current);
            return true;
        } else {
            current = (current + 1) < table->capacity ? (current + 1) : 0;
        }
    }

    if (current != end) {
        OBEL_SAFE_DEREF_ASSIGN(i, current);
    }

    return false;
}

// MARK: - Sparse Coordinate Grid

OBEL_COORD_FUNC obel_coord_grid_t obel_coord_grid_make(float stride, obel_aabb_t aabb) {
    obel_coord_t max_coord = obel_coord_grid_bin(stride, aabb.min, aabb.max);
    return (obel_coord_grid_t){
        .origin = aabb.min,
        .stride = stride,
        .width = max_coord.x + 1,
        .height = max_coord.y + 1,
        .depth = max_coord.z + 1
    };
}

OBEL_COORD_FUNC obel_coord_t obel_coord_grid_bin(float stride, obel_vector3_t origin, obel_vector3_t point) {
    obel_vector3_t coord = obel_vector_div(obel_vector_sub(point, origin), stride);
    return obel_coord_make(floorf(coord.x), floorf(coord.y), floorf(coord.z));
}

OBEL_COORD_FUNC obel_coord_t obel_coord_grid_bin(obel_coord_grid_t *grid, obel_vector3_t point) {
    return obel_coord_grid_bin(grid->stride, grid->origin, point);
}

// MARK: - Coordinate Heap

OBEL_COORD_FUNC uint32_t obel_coord_heap_insert(obel_coord_heap_t *heap, obel_coord_t coord) {
    if (heap->count == heap->capacity) {
        return 0;
    }

    uint32_t current = ++heap->count;
    heap->coords[current] = coord;

    if (current == OBEL_HEAP_ROOT) {
        return current;
    }

    uint32_t parent = current;
    switch (heap->order) {
        case obel_heap_order_min: {
            while ((heap->coords[current].bits & heap->mask) <
                   (heap->coords[parent = OBEL_HEAP_PARENT(current)].bits & heap->mask)) {
                obel_swap(&heap->coords[current].bits, &heap->coords[parent].bits);
                current = parent;
            }
            break;
        }

        case obel_heap_order_max: {
            while ((heap->coords[current].bits & heap->mask) >
                   (heap->coords[parent = OBEL_HEAP_PARENT(current)].bits & heap->mask)) {
                obel_swap(&heap->coords[current].bits, &heap->coords[parent].bits);
                current = parent;
            }
            break;
        }
    }

    return current;
}

OBEL_COORD_FUNC obel_coord_t obel_coord_heap_remove(obel_coord_heap_t *heap, uint32_t i) {
    obel_coord_t coord = heap->coords[i];
    heap->coords[i] = heap->coords[heap->count];
    heap->coords[heap->count--] = (obel_coord_t){ 0 };

    uint32_t left, right = 0;
    uint32_t current = i;

    switch (heap->order) {
        case obel_heap_order_min: {
            uint32_t smallest = 0;
            while ((smallest = current) < heap->count) {
                if ((heap->coords[left = OBEL_HEAP_LEFT(current)].bits & heap->mask) <
                    (heap->coords[smallest].bits & heap->mask)) {
                    if (left <= heap->count) {
                        smallest = left;
                    }
                }

                if ((heap->coords[right = OBEL_HEAP_RIGHT(current)].bits & heap->mask) <
                    (heap->coords[smallest].bits & heap->mask)) {
                    if (right <= heap->count) {
                        smallest = right;
                    }
                }

                if (current == smallest) {
                    break;
                }

                obel_swap(&heap->coords[current].bits, &heap->coords[smallest].bits);
                current = smallest;
            }
            break;
        }

        case obel_heap_order_max: {
            uint32_t largest = 0;
            while ((largest = current) < heap->count) {
                if ((heap->coords[left = OBEL_HEAP_LEFT(current)].bits & heap->mask) >
                    (heap->coords[largest].bits & heap->mask)) {
                    if (left <= heap->count) {
                        largest = left;
                    }
                }

                if ((heap->coords[right = OBEL_HEAP_RIGHT(current)].bits & heap->mask) >
                    (heap->coords[largest].bits & heap->mask)) {
                    if (right <= heap->count) {
                        largest = right;
                    }
                }

                if (current == largest) {
                    break;
                }

                obel_swap(&heap->coords[current].bits, &heap->coords[largest].bits);
                current = largest;
            }
            break;
        }
    }

    return coord;
}

OBEL_COORD_FUNC bool obel_coord_heap_remove(obel_coord_heap_t *heap, obel_coord_t coord) {
    for (uint32_t i = OBEL_HEAP_ROOT; i < heap->count; i++) {
        if (heap->coords[i].x == coord.x && heap->coords[i].y == coord.y && heap->coords[i].z == coord.z) {
            obel_coord_heap_remove(heap, i);
            return true;
        }
    }

    return false;
}

// MARK: - Coordinate Marching

OBEL_COORD_FUNC uint32_t obel_coord_march(obel_coord_marcher_t *marcher) {
    obel_coord_t *coords = marcher->band.table.coords;

    // Extract the root vertex from the max heap.
    obel_coord_t point_coord = obel_coord_heap_remove(&marcher->vertices, OBEL_HEAP_ROOT);

    // Reset the arrival time to zero.
    uint32_t point_coord_i;
    assert(obel_coord_table_find(&marcher->band.table, point_coord, &point_coord_i));
    coords[point_coord_i].bits &= ~OBEL_COORD_BITS_ARR_TIME;

    // Mark the vertex as alive.
    coords[point_coord_i].bits &= ~OBEL_COORD_BITS_CLOSE;
    coords[point_coord_i].bits |= OBEL_COORD_BITS_ALIVE;

    // Even though the vertex is not considered CLOSE anymore, insert it into the propagation heap
    // so its neighbors will be visited.
    obel_coord_heap_insert(&marcher->propagation, point_coord);

    // Propagate until we have no more CLOSE coordinates to visit.
    obel_coord_propagate(marcher);

    return point_coord_i;
}

OBEL_COORD_FUNC void obel_coord_propagate(obel_coord_marcher_t *marcher) {
    obel_coord_t *coords = marcher->band.table.coords;

    while (marcher->propagation.count > 0) {
        obel_coord_t trial_coord = obel_coord_heap_remove(&marcher->propagation, OBEL_HEAP_ROOT);
        uint16_t trial_arrival_time = trial_coord.bits & OBEL_COORD_BITS_ARR_TIME;

        uint32_t trial_i = 0;
        obel_coord_table_find(&marcher->band.table, trial_coord, &trial_i);
        coords[trial_i].bits &= ~OBEL_COORD_BITS_CLOSE;
        coords[trial_i].bits |= OBEL_COORD_BITS_ALIVE;

        uint32_t neighbor_i = 0;
        obel_coord_t neighbors[6] = {
            obel_coord_make(trial_coord.x + 1, trial_coord.y, trial_coord.z),
            obel_coord_make(trial_coord.x - 1, trial_coord.y, trial_coord.z),
            obel_coord_make(trial_coord.x, trial_coord.y + 1, trial_coord.z),
            obel_coord_make(trial_coord.x, trial_coord.y - 1, trial_coord.z),
            obel_coord_make(trial_coord.x, trial_coord.y, trial_coord.z + 1),
            obel_coord_make(trial_coord.x, trial_coord.y, trial_coord.z - 1)
        };

        for (uint32_t i = 0; i < 6; i++) {
            // Ensure the neighbor exists inside the offset band.
            if (obel_coord_table_find(&marcher->band.table, neighbors[i], &neighbor_i)) {
                if (coords[neighbor_i].bits & OBEL_COORD_BITS_ALIVE) {
                    continue;
                }

                uint16_t neighbor_arrival_time = coords[neighbor_i].bits & OBEL_COORD_BITS_ARR_TIME;

                // If the neighbor's arrival time is less than the trial's arrival time, we have
                // collided with another Voronoi cell, so stop propagation in this "direction".
                if (neighbor_arrival_time < trial_arrival_time) {
                    // The neighbor is now visited by another wave.
                    coords[neighbor_i].bits += OBEL_COORD_BITS_PROP_WAVES_ONE;

                    // The number of propagation waves is three (both bits set), we've reached
                    // another Voronoi vertex, which could become the next point coordinate.
                    if (OBEL_COORD_IS_VORONOI(coords[neighbor_i].bits)) {
                        obel_coord_heap_insert(&marcher->vertices, coords[neighbor_i]);
                    }
                } else {
                    // Increment the arrival time, using the trial coordinate's arrival time.
                    coords[neighbor_i].bits = ((coords[neighbor_i].bits & ~OBEL_COORD_BITS_ARR_TIME) |
                                               ((trial_coord.bits & OBEL_COORD_BITS_ARR_TIME) + 1));

                    if (OBEL_COORD_IS_VORONOI(coords[neighbor_i].bits)) {
                        obel_coord_heap_remove(&marcher->vertices, coords[neighbor_i]);
                    }

                    // The neighbor coordinate has now only been visited by a single propagation
                    // wave.
                    coords[neighbor_i].bits &= ~OBEL_COORD_BITS_PROP_WAVES;
                    coords[neighbor_i].bits |= OBEL_COORD_BITS_PROP_WAVES_ONE;

                    // Insert the neighbor coordinate into the propagation wave.
                    coords[neighbor_i].bits |= OBEL_COORD_BITS_CLOSE;
                    obel_coord_heap_insert(&marcher->propagation, coords[neighbor_i]);
                }
            }
        }
    }
}
